File: module2/class1/exemplo2.cs
/*
================================================================================
ATIVIDADE PRÁTICA 2 - IDENTIFICAÇÃO DE VAZAMENTO DE MEMÓRIA COM HEAP SNAPSHOTS (C#)
================================================================================

OBJETIVO:
- Criar aplicação que faça alocações não-gerenciadas sem liberar memória
- Capturar snapshots de uso da memória com Memory Usage Tool
- Comparar snapshots para identificar crescimento anormal do heap
- Refatorar código para liberar memória corretamente
- Repetir profiling para validar correção

PROBLEMA:
- Funções fazem Marshal.AllocHGlobal mas nunca fazem FreeHGlobal
- Memory Usage Tool mostrará crescimento contínuo do heap não-gerenciado
- Cada iteração adiciona ~600KB sem liberar

SOLUÇÃO:
- Adicionar Marshal.FreeHGlobal para cada AllocHGlobal
- Resultado: estabilização do uso de memória não-gerenciada

================================================================================
*/

using System;
using System.Collections.Generic;
using System.Runtime.InteropServices;
using System.Threading;

class Program {
    private static List<IntPtr> leakedPointers = new List<IntPtr>();
    
    static void LeakMemory() {
        IntPtr leak = Marshal.AllocHGlobal(50000 * sizeof(int)); // MEMORY LEAK: Missing Marshal.FreeHGlobal(leak); - Add this line to fix the leak
        leakedPointers.Add(leak);
        
        unsafe {
            int* ptr = (int*)leak.ToPointer();
            for (int i = 0; i < 50000; i++) {
                ptr[i] = i * i;
            }
        }
    }
    
    static void CreateLargeLeak() {
        IntPtr bigLeak = Marshal.AllocHGlobal(25000 * sizeof(double)); // MEMORY LEAK: Missing Marshal.FreeHGlobal(bigLeak); - Add this line to fix the leak
        leakedPointers.Add(bigLeak);
        
        unsafe {
            double* ptr = (double*)bigLeak.ToPointer();
            for (int i = 0; i < 25000; i++) {
                ptr[i] = i * 3.14159;
            }
        }
    }
    
    static void Main() {
        Console.WriteLine("Starting memory leak demonstration...");
        Console.WriteLine("Take heap snapshots at different iterations to see memory growth");
        
        for (int i = 0; i < 100; i++) {
            LeakMemory();
            CreateLargeLeak();
            
            Console.WriteLine($"Iteration: {i + 1} - Heap should be growing...");
            
            if ((i + 1) % 20 == 0) {
                Console.WriteLine($"*** GOOD POINT FOR HEAP SNAPSHOT *** (Iteration {i + 1})");
                Thread.Sleep(2000);
            }
            
            Thread.Sleep(100);
        }
        
        Console.WriteLine("Program finished - memory was never released!");
        
        // TO FIX THE LEAKS, UNCOMMENT THE LINES BELOW:
        // foreach (IntPtr ptr in leakedPointers) {
        //     Marshal.FreeHGlobal(ptr);
        // }
    }
}

/*
================================================================================
VERSÃO CORRIGIDA (DESCOMENTE PARA USAR A VERSÃO SEM VAZAMENTOS)
================================================================================

using System;
using System.Collections.Generic;
using System.Runtime.InteropServices;
using System.Threading;

class Program {
    static void UseMemoryCorrectly() {
        IntPtr data = Marshal.AllocHGlobal(50000 * sizeof(int));
        
        unsafe {
            int* ptr = (int*)data.ToPointer();
            for (int i = 0; i < 50000; i++) {
                ptr[i] = i * i;
            }
        }
        
        // CORREÇÃO: Liberar memória não-gerenciada
        Marshal.FreeHGlobal(data);
    }
    
    static void CreateLargeDataCorrectly() {
        IntPtr bigData = Marshal.AllocHGlobal(25000 * sizeof(double));
        
        unsafe {
            double* ptr = (double*)bigData.ToPointer();
            for (int i = 0; i < 25000; i++) {
                ptr[i] = i * 3.14159;
            }
        }
        
        // CORREÇÃO: Liberar memória não-gerenciada
        Marshal.FreeHGlobal(bigData);
    }
    
    static void Main() {
        Console.WriteLine("Starting corrected memory management demonstration...");
        Console.WriteLine("Take heap snapshots - unmanaged memory usage should remain stable");
        
        for (int i = 0; i < 100; i++) {
            UseMemoryCorrectly();
            CreateLargeDataCorrectly();
            
            Console.WriteLine($"Iteration: {i + 1} - Heap should be stable...");
            
            if ((i + 1) % 20 == 0) {
                Console.WriteLine($"*** GOOD POINT FOR HEAP SNAPSHOT *** (Iteration {i + 1})");
                Thread.Sleep(2000);
            }
            
            Thread.Sleep(100);
        }
        
        Console.WriteLine("Program finished - all memory was properly released!");
    }
}

================================================================================
*/

File: module2/class1/exemplo2.cpp
/*
================================================================================
ATIVIDADE PRÁTICA 2 - IDENTIFICAÇÃO DE VAZAMENTO DE MEMÓRIA COM HEAP SNAPSHOTS (C++)
================================================================================

OBJETIVO:
- Criar aplicação que faça alocações dinâmicas sem liberar memória
- Capturar snapshots de uso da memória com Memory Usage Tool
- Comparar snapshots para identificar crescimento anormal do heap
- Refatorar código para liberar memória corretamente
- Repetir profiling para validar correção

PROBLEMA:
- Funções fazem new[] mas nunca fazem delete[]
- Memory Usage Tool mostrará crescimento contínuo do heap
- Cada iteração adiciona ~600KB sem liberar

SOLUÇÃO:
- Adicionar delete[] para cada new[]
- Resultado: estabilização do uso de memória

================================================================================
*/

#include <iostream>
#include <chrono>
#include <thread>
using namespace std;

void leak_memory() {
    int* leak = new int[50000]; // MEMORY LEAK: Missing delete[] leak; - Add this line to fix the leak
    for (int i = 0; i < 50000; ++i) {
        leak[i] = i * i;
    }
}

void create_large_leak() {
    double* big_leak = new double[25000]; // MEMORY LEAK: Missing delete[] big_leak; - Add this line to fix the leak
    for (int i = 0; i < 25000; ++i) {
        big_leak[i] = i * 3.14159;
    }
}

int main() {
    cout << "Starting memory leak demonstration..." << endl;
    cout << "Take heap snapshots at different iterations to see memory growth" << endl;
    
    for (int i = 0; i < 100; ++i) {
        leak_memory();
        create_large_leak();
        
        cout << "Iteration: " << i + 1 << " - Heap should be growing..." << endl;
        
        if ((i + 1) % 20 == 0) {
            cout << "*** GOOD POINT FOR HEAP SNAPSHOT *** (Iteration " << i + 1 << ")" << endl;
            this_thread::sleep_for(chrono::milliseconds(2000));
        }
        
        this_thread::sleep_for(chrono::milliseconds(100));
    }
    
    cout << "Program finished - memory was never released!" << endl;
    return 0;
}

/*
================================================================================
VERSÃO CORRIGIDA (DESCOMENTE PARA USAR A VERSÃO SEM VAZAMENTOS)
================================================================================

#include <iostream>
#include <chrono>
#include <thread>
using namespace std;

void use_memory_correctly() {
    int* data = new int[50000];
    for (int i = 0; i < 50000; ++i) {
        data[i] = i * i;
    }
    // CORREÇÃO: Liberar memória alocada
    delete[] data;
}

void create_large_data() {
    double* big_data = new double[25000];
    for (int i = 0; i < 25000; ++i) {
        big_data[i] = i * 3.14159;
    }
    // CORREÇÃO: Liberar memória alocada
    delete[] big_data;
}

int main() {
    cout << "Starting corrected memory management demonstration..." << endl;
    cout << "Take heap snapshots - memory usage should remain stable" << endl;
    
    for (int i = 0; i < 100; ++i) {
        use_memory_correctly();
        create_large_data();
        
        cout << "Iteration: " << i + 1 << " - Heap should be stable..." << endl;
        
        if ((i + 1) % 20 == 0) {
            cout << "*** GOOD POINT FOR HEAP SNAPSHOT *** (Iteration " << i + 1 << ")" << endl;
            this_thread::sleep_for(chrono::milliseconds(2000));
        }
        
        this_thread::sleep_for(chrono::milliseconds(100));
    }
    
    cout << "Program finished - all memory was properly released!" << endl;
    return 0;
}

================================================================================
*/

